### PR TITLE
[move-prover] a basic stackless bytecode interpreter

### PR DESCRIPTION
**Pull request by mengxu-fb**
This PR is the base PR for the stackless bytecode interpreter.

The first four commits are the major ones in the PR:

- the first commit presents an interpreter for all stackless bytecode and
operations, except native functions (including emit event). In theory,
if the implementation is correct, the interpreter itself will be able to
execute all Move programs that do not have native functions successfully.
However, the assertions are not interpreted yet because the assertions
are in expressions, which is not part of the stackless bytecode. Places to
handle expressions and native functions are marked with TODOs in the
code. Those items will be handled in subsequent PRs.

- the second commit integrates the stackless vm as an option we can pass
to the exciting Move unit testing framework (from @tzakian).
```bash
cargo run move-unit-test -- --stackless <path-to-move-sources>
```
will instruct the unit test framework to use the stackless interpreter to
execute the unit test cases, instead of the original Move VM. In theory,
the stackless VM and the Move VM should present exactly the same results
(unless some assertions are violated, which can only be caught by the
stackless VM). This property is not guaranteed in this PR as of now, but
we will have fairly good confidence once we have all the unit tests for the
stackless bytecode.

- the third commit starts the testsuite for the stackless VM to check it
interprets each bytecode in with the correct semantics. Each test case
is a Move program with `#[test]` features and the program will be executed
by both the Move VM and stackless VM for cross-comparision. Currently
the testsuite has only a nop test to make sure the overall flow works. A lot
more test cases will be added soon.

- the fourth commit adds the `-v, --verbose` flag to the unit testing framework
as well as the stackless VM to allow debug messages to be printed.

Starting from the fifth commit onwards, the commit content is about adding
unit test cases for each stackless bytecode and fix the issues found during the
testing.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

CI + new unit test cases

### CODE REVIEW

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

File: Cargo.toml
Comment (wrwg) [inline Cargo.toml:93]: Why in its own crate?

File: language/move-prover/interpreter/src/concrete/local_state.rs
Comment (wrwg) [inline language/move-prover/interpreter/src/concrete/local_state.rs:110]: Don't you get this with `debug_assert!`?

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [inline language/move-prover/interpreter/src/concrete/player.rs:76]: It's not really step-by-step on bytecode level but on function invocation level?

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [inline language/move-prover/interpreter/src/concrete/player.rs:155]: nit: perhaps `use Bytecode::*` for brevity

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [inline language/move-prover/interpreter/src/concrete/player.rs:150]: Wondering why you don't make this functions of FunctionContext, or create a `struct Player { ctxt: ...}` or similar.

Static functions like this are a difficult to extend and maintain.

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [inline language/move-prover/interpreter/src/concrete/player.rs:205]: I think Store is a leftover from old days and really doesn't differ from Copy (or Move!). I believe its generated if something is popped from the stack and stored in a temp. 

Whether its Move or Copy (if this is relevant here) is best seen in the borrow analysis in bytecode.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [inline language/move-prover/interpreter/src/concrete/player.rs:366]: is_mut should be always true.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [inline language/move-prover/interpreter/src/concrete/player.rs:896]: Add a TODO here that this needs to be extended to check for actual address in borrow graph. There is a bug here right now and there is already a TODO in bytcodegenerator (fix is coming up in a while).

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [inline language/move-prover/interpreter/src/concrete/player.rs:909]: This should not be an assert. Rather it must be a test whether root_idx == local_root and only if it succeeds we perform the update.

Consider

```
let a = 1; let b = 2;
let r = if (b) { &mut a } else  {&mut b };
*r = 3;
```

We will create two WriteBacks:

```
write_back[Local(a)](r)
write_back[Local(b)](r)
```
Only one of them will be effective, depending on what we find Pointer::Local.

A similar logic is currently missing for global write-backs in the borrow graph and therefore in the generated bytecode, therefore the TODO I mentioned above.

**Review by wrwg — COMMENTED** [review]
This looks promising! Initial comments.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [reply language/move-prover/interpreter/src/concrete/player.rs:76]: > It's not really step-by-step on bytecode level but on function invocation level?

Seems not step-by-step at all for now. We can consider this good to go for the first version, but eventually, we may want an executor which can really run step-by-step.

File: Cargo.toml
Comment (mengxu-fb) [reply Cargo.toml:93]: To avoid cyclic dependency.

`move-unit-test` depends on `move-prover/interpreter`, therefore, in `move-prover/interpreter`, we cannot depend on `move-unit-test` to run the tests

File: language/move-prover/interpreter/src/concrete/local_state.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/local_state.rs:110]: Yes, but I used `cfg!(debug_assertions)` consistently in the code to make `grep` a bit easier. Same effect with `debug_assert`.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs:76]: The `exec_bytecode()` is the `step-by-step` part, or more precisely, `bytecode-by-bytecode`.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs:155]: I am not a big fan of the `::*` notion as well as placing imports inside the function. But I do see the brevity, so, it is a hard choice.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs:150]: That is a good point! Honestly, I am not sure why I ended up coding it in a functional way instead of wrapping them around a `struct Player { ctxt: ...}`.

One reason I can think of now is that I want to re-use many of the `handle_*` functions when evaluating expressions, and there are several `handle_*` that does not require a `FunctionContext` to work (e.g., the arithmetics and write-backs). Having them inside a `FunctionContext` might make the expression evaluator a bit hard to work.

But that is just my consideration. I might be just overthinking on this.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs:205]: I see `Store` is treated like `Move` everywhere in the pipeline (and in boogie translator as well), but when executing it, I found that `dst := Store(src)` does not clear the local slot at `src`. So I treated it as a `Copy` here to keep things going. I'll come back with a better handling and remove the TODO.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs:366]: At the end of the transformation, yes, `is_mut` should always be true, but before the eliminate-imm-ref passs, I think either mut or imm refs can be created?

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs:909]: Ah, I see. Thanks for the clarification! Will update.

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs:366]: Nvm, I found that it does not make sense to execute the stackless bytecode after arbitrary passes in the pipeline. The program might not be correct. For example:
- after `ReachDefAnalysis`, we may have two `$t2 = move($t1); $t3 = move($t1);`
- before write-backs are instrumented, the modified local root is never updated...

So I removed the code that was intended for per-pass processing and only execute at the end of the transformation pipeline.

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

File: Cargo.toml
Comment (wrwg) [reply Cargo.toml:93]: In fact if crate A depends on crate B, then still the B testsuite can depend on A. Rust breaks the cycle for tests. We have examples for this in the prover, e.g. docgen depends on the prover main crate, and docgen tests depend on the prover crate. It's not that important though.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [reply language/move-prover/interpreter/src/concrete/player.rs:76]: A real step-by-step execution would provide a public entrypoint `fn exec_next_step(&mut State)` or similar which would execute the next bytecode instruction and then return. This would allow single-stepping through execution and other advanced instrumentation scenarios. I'm not saying we need this now, just confused because you mentioned stepping. If I do not misunderstand, what you do right now is not single-stepping in this sense.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (wrwg) [reply language/move-prover/interpreter/src/concrete/player.rs:150]: > there are several handle_* that does not require a FunctionContext to work

That is exactly the problem with this design approach. Maybe they do not need the FunctionContext now, but then tomorrow you need to make a change where they do, and now you may be faced with cascading changes because you add a parameter which needs to be added at all caller sites (and caller caller sites and so on). This is what makes this design so bad for maintainability and extensibility. I think being able to call handle_X is a minor consideration in this context, you can also construct a temporary context if this really would be needed. I would recommend to do this change now before the code grows further.

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

File: language/tools/move-unit-test/src/test_runner.rs
Comment (tzakian) [inline language/tools/move-unit-test/src/test_runner.rs]: Is there any way of getting the final storage state out of this after execution? Eventually we'll want to support printing out global states when a test fails. It's fine if it's not supported, but would be interesting if we could this if e.g., the stackless and bytecode tests differ.

File: language/tools/move-unit-test/src/test_runner.rs
Comment (tzakian) [inline language/tools/move-unit-test/src/test_runner.rs:175]: Could you add a TODO in the `execute_via_move_vm` function to support collecting VM logs if the `verbose` flag is set as well?

File: language/tools/move-unit-test/src/test_runner.rs
Comment (tzakian) [inline language/tools/move-unit-test/src/test_runner.rs:315]: I'm thinking we probably want to add in a different error type to represent these types of errors. In my mind when I was implementing this, it was a way of saying "this needs to abort, but I don't care with what abort code" which is different from  "this should fail with something other than an abort code."

What would you think about adding in a new failure type for this? e.g., 
```
#[expected_failure(abort)] // this should abort but I don't care about the code
#[expected_failure(abort = <code>)] // this should abort with <code>
#[expected_failure(internal)] // this should fail with an internal error (i.e., any error other than abort or OUT_OF_GAS)
```

It's fine to keep this as-is for now, but could you possibly file an issue for this, and you or I can add it in after this PR lands.

File: language/move-prover/interpreter-testsuite/tests/concrete_check/global_basics.move
Comment (tzakian) [inline language/move-prover/interpreter-testsuite/tests/concrete_check/global_basics.move:70]: I'm wondering, do you want to assert on these exists checks?

File: language/move-prover/interpreter-testsuite/tests/concrete_check_testsuite.rs
Comment (tzakian) [inline language/move-prover/interpreter-testsuite/tests/concrete_check_testsuite.rs:19]: `use_stackless_vm` doesn't seem quite right with this change -- it makes me think of only using the stackless VM for execution, and less about comparing execution results between the two. 

What would you think of changing the name of this to something like `compare_against_stackless_vm`, `check_stackless_vm`, or something else along these lines?

File: language/tools/move-unit-test/src/test_runner.rs
Comment (mengxu-fb) [reply language/tools/move-unit-test/src/test_runner.rs]: Yes, we can have the two `execute_via_*` functions return a tuple `(VMResult<Vec<Vec<u8>>>, InMemoryStorage)`, which is easy to do for both the Move VM version and the stackless VM version.

File: language/tools/move-unit-test/src/test_runner.rs
Comment (mengxu-fb) [reply language/tools/move-unit-test/src/test_runner.rs:315]: I like the `#[expected_failure(internal)]` failure type for this. I'll file an issue.

File: language/move-prover/interpreter-testsuite/tests/concrete_check/global_basics.move
Comment (mengxu-fb) [reply language/move-prover/interpreter-testsuite/tests/concrete_check/global_basics.move:70]: I guess either way works? The main goal of this testsuite is to check that the stackless VM behaves like the Move VM so I am "implicitly" relying on the return value of the function (and the cross-vm comparison) for checking.

File: language/move-prover/interpreter-testsuite/tests/concrete_check_testsuite.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter-testsuite/tests/concrete_check_testsuite.rs:19]: Right! It was originally named to represent only using the stackless VM but later I changed its meaning without the name :(. I'll rename it to `check_stackless_vm`.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs:76]: Ah, I see. But I still feel that the `exec_bytecode` function is implemented with this spirit.
```
fn exec_bytecode(
    ctxt: &FunctionContext,
    bytecode: &Bytecode,
    local_state: &mut LocalState,
    global_state: &mut GlobalState,
) -> PartialVMResult<()> {}
```
It takes a bytecode, and two mutable states (local and global) and just execute that bytecode and then return.

File: language/tools/move-unit-test/src/test_runner.rs
Comment (tzakian) [reply language/tools/move-unit-test/src/test_runner.rs]: Having it return a `ChangeSet` at the end would be the best if possible as that would fit in the most with my current implementation of this I'm working on, but I realize that it very well might not be. Either way, this isn't something that I think needs to be done in this PR and can be be done later on.

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs:150]: > I would recommend to do this change now before the code grows further.

Totally agree. I'll make the change now.

File: language/move-prover/interpreter-testsuite/tests/concrete_check/global_basics.move
Comment (tzakian) [reply language/move-prover/interpreter-testsuite/tests/concrete_check/global_basics.move:70]: Oh, woah..that's really cool actually! I think leaving this as-is then is fine

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

File: language/move-prover/interpreter/src/concrete/player.rs
Comment (mengxu-fb) [reply language/move-prover/interpreter/src/concrete/player.rs]: Updated. Now everything wraps around the `FunctionContext` and if later `handle_X` needs something in the context, we don't need the cascading changes of APIs.

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A bytecode-interpreter 0.1.0 (workspace, path 'language/move-prover/interpreter')
    * features: [none]
```

**Review by wrwg — APPROVED** [review]
Approving so we can iterate. Consider also getting Tim to be OK with it.

**Review by tzakian — APPROVED** [review]
LGTM, nothing on my side that's blocking this going in :)

**Comment by mengxu-fb** [comment]
Thank you @wrwg and @tzakian for the feedbacks! Will further iterate from here.

/land

**Pull request merged**
